hogben/models/base.py

```diff
@@ -419,3 +443,88 @@ def nested_sampling(self,
+
+
+class MagneticSLD(refnx.reflect.structure.Slab):
+    """
+    A class to represent a layer with a magnetic SLD component.
+
+    This class extends the `Slab` class from `refnx.reflect.structure` to
+    include properties for magnetic Scattering Length Density (SLD) as well.
+
+    Attributes:
+        SLD_n (float): Nuclear scattering length density.
+        SLD_m (float): Magnetic scattering length density.
+        thickness (float): Thickness of the layer.
+        roughness (float): Roughness of the layer.
+        name (str): Name of the layer.
+    """
+
+    def __init__(self,
+                 SLDn: float = 0,
+                 SLDm: float = 0,
+                 thick: float = 0,
+                 rough: float = 0,
+                 name: str = 'Magnetic Layer'):
+        """
+        Initialize a MagneticSLD object.
+
+        Parameters:
+            SLDn (float): Nuclear scattering length density. Default is 0.
+            SLDm (float): Magnetic scattering length density. Default is 0.
+            thick (float): Thickness of the layer. Default is 0.
+            rough (float): Roughness of the layer. Default is 0.
+            name (str): Name of the layer. Default is "Magnetic Layer".
+        """
+        self.SLD_n = SLDn
+        self.SLD_m = SLDm
+        self.thickness = thick
+        self.roughness = rough
+        self.name = name
+        super().__init__(thick=self.thickness, sld=self.SLD_n,
+                         rough=self.roughness, name=self.name)
+
+    @property
+    def spin_up(self):
+        """
+        Calculate the spin-up scattering length density.
+
+        Returns:
+            SLD: An SLD object representing the spin-up component with the
+            appropriate thickness and roughness.
+        """
+        SLD_value = self.SLD_n + self.SLD_m
+        return SLD(SLD_value, name='spin_up')(thick=self.thickness,
+                                              rough=self.rough)
+
+    @property
+    def spin_down(self):
+        """
+        Calculate the spin-down scattering length density.
+
+        Returns:
+            SLD: An SLD object representing the spin-down component with the
+            appropriate thickness and roughness.
+        """
+        SLD_value = self.SLD_n - self.SLD_m
+        return SLD(SLD_value, name='spin_down')(thick=self.thickness,
+                                                rough=self.rough)
+
+    def __call__(self, thick=None, rough=None):
+        """
+        Update the thickness and roughness of the layer.
+
+        Parameters:
+            thick (float, optional): New thickness of the layer. If None, the
+            current thickness is retained.
+            rough (float, optional): New roughness of the layer. If None, the
+            current roughness is retained.
+
+        Returns:
+            MagneticSLD: The updated MagneticSLD object.
+        """
+        self.thickness = thick if thick is not None else self.thickness
```

Comment:
This dunder method is called when calling the object. So it's the second pair of parenthesis in: `MagneticSLD(ref_sld, ref_msld, name="Layer 3")(thick=ref_thick, rough=0)`

The idea is so that it's basically identical to `refnx` layers, where a layer is created as:
`SLD(ref_sld, name="Layer 3")(thick=ref_thick, rough=0)`.

This dunder method is a bit of a hack, but it makes it possible to simply replace `SLD` with `MagneticSLD`, and it should remain 100% compatible without issues (given the `SLDm` attribute is optional and defaults to 0). Then one can just add the second `SLDm` attribute (making it `MagneticSLD(ref_sld, ref_msld, name="Layer 3")(thick=ref_thick, rough=0)`) for a magnetic layer.

As far as I'm aware, the SLD cannot be given upon calling the `SLD` object in `refnx`. Just the thickness and roughness. Hence only having those two there.

Reply:
Ahh, I get it. Thanks, as it is is fine then!